pfcpiface/bess.go

```diff
@@ -985,6 +1005,109 @@ func (b *bess) delCounter(ctx context.Context, done chan<- bool, ctrID uint32, c
 	}()
 }
 
+func (b *bess) processSliceMeter(ctx context.Context, any *anypb.Any, method upfMsgType) {
+	if method != upfMsgTypeAdd && method != upfMsgTypeDel && method != upfMsgTypeClear {
+		log.Errorln("Invalid method name: ", method)
+		return
+	}
+
+	methods := [...]string{"add", "add", "delete", "clear"}
+
+	_, err := b.client.ModuleCommand(
+		ctx, &pb.CommandRequest{
+			Name: "sliceMeter",
+			Cmd:  methods[method],
+			Arg:  any,
+		},
+	)
+	if err != nil {
+		log.Errorln("sliceMeter method failed!:", err)
+	}
+}
+
+func (b *bess) addSliceMeter(ctx context.Context, done chan<- bool, meterConfig SliceMeterConfig) {
+	go func() {
+		var (
+			any                           *anypb.Any
+			err                           error
+			cir, pir, cbs, ebs, pbs, gate uint64
+		)
+
+		// Uplink slice meter config
+		if meterConfig.UplinkRateBps != 0 {
+			gate = sliceMeterGateMeter
+			cir = 1                             // Mark all traffic as yellow
+			pir = meterConfig.UplinkRateBps / 8 // bit/s to byte/s
+		} else {
+			gate = sliceMeterGateUnmeter
+		}
+		if meterConfig.UplinkBurstBytes != 0 {
+			cbs = 1 // Mark all traffic as yellow
+			pbs = meterConfig.UplinkBurstBytes
+			ebs = 0 // Unused
+		} else {
+			cbs = DefaultBurstSize
+			pbs = DefaultBurstSize
+			ebs = 0 // Unused
+		}
+		q := &pb.QosCommandAddArg{
+			Gate:                    gate,
+			Cir:                     cir, /* committed info rate */
+			Pir:                     pir, /* peak info rate */
+			Cbs:                     cbs, /* committed burst size */
+			Pbs:                     pbs, /* Peak burst size */
+			Ebs:                     ebs, /* Excess burst size */
+			AdjustMeterPacketLength: -14, /* Exclude Ethernet header */
```

Comment:
This uplink meter is counting packets slightly wrong, as there is both N6 (encaped) and N9 (not encaped) traffic. Using the "stricter" `-14` ensures that the overall limit is met, albeit limiting N9 UE traffic a bit more than configured.

Reply:
As mentioned in the above https://github.com/omec-project/upf-epc/pull/324#discussion_r705514543, we don't need to worry about N9 for now,  so it should be ok to always subtract -50

Reply:
Good to know!

Shouldn't it be `-14` for N6, though? Non-encapped means just one Ethernet header (14 bytes) to ignore.

Reply:
Right, because we do metering after decap... right? I guess I was confused by your comment about -14 not being the ideal value for N6, but maybe I'm reading that wrong.